src/Razor/Microsoft.AspNetCore.Razor.Language/test/IntegrationTests/ComponentCodeGenerationTestBase.cs

```diff
@@ -3152,6 +3152,391 @@ public class MyComponent<TItem> : ComponentBase
+using Microsoft.AspNetCore.Components;
+
+namespace Test
+{
+    [CascadingTypeParameter(nameof(TKey))]
+    [CascadingTypeParameter(nameof(TValue))]
+    [CascadingTypeParameter(nameof(TOther))]
+    public class Parent<TKey, TValue, TOther> : ComponentBase
+    {
+        [Parameter] public Dictionary<TKey, TValue> Data { get; set; }
+        [Parameter] public TOther Other { get; set; }
+        [Parameter] public RenderFragment ChildContent { get; set; }
+    }
+
+    public class Child<TOther, TValue, TKey, TChildOnly> : ComponentBase
+    {
+        [Parameter] public ICollection<TChildOnly> ChildOnlyItems { get; set; }
+    }
+}
+"));
+
+            // Act
+            var generated = CompileToCSharp(@"
+<Parent Data=""@(new System.Collections.Generic.Dictionary<int, string>())"" Other=""@DateTime.MinValue"">
+    <Child ChildOnlyItems=""@(new[] { 'a', 'b', 'c' })"" />
+</Parent>");
+
+            // Assert
+            AssertDocumentNodeMatchesBaseline(generated.CodeDocument);
+            AssertCSharpDocumentMatchesBaseline(generated.CodeDocument);
+            CompileToAssembly(generated);
+        }
+
+        [Fact]
+        public void ChildComponent_Generic_TypeInference_CascadedWithUnrelatedGenericType_CreatesDiagnostic()
+        {
+            // It would succeed if you changed this to Column<TItem, TUnrelated>, or if the Grid took a parameter
+            // whose type included TItem and not TUnrelated. It just doesn't work if the only inference parameters
+            // also include unrelated generic types, because the inference methods we generate don't know what
+            // to do with extra type parameters. It would be nice just to ignore them, but at the very least we
+            // have to rewrite their names to avoid clashes and figure out whether multiple unrelated generic
+            // types with the same name should be rewritten to the same name or unique names.
+
+            // Arrange
+            AdditionalSyntaxTrees.Add(Parse(@"
+using Microsoft.AspNetCore.Components;
+
+namespace Test
+{
+    [CascadingTypeParameter(nameof(TItem))]
+    [CascadingTypeParameter(nameof(TUnrelated))]
+    public class Grid<TItem, TUnrelated> : ComponentBase
+    {
+        [Parameter] public System.Collections.Generic.Dictionary<TItem, TUnrelated> Items { get; set; }
+        [Parameter] public RenderFragment ChildContent { get; set; }
+    }
+
+    public class Column<TItem> : ComponentBase
+    {
+    }
+}
+"));
+
+            // Act
+            var generated = CompileToCSharp(@"
+<Grid Items=""@(new Dictionary<int, string>())""><Column /></Grid>");
+
+            // Assert
+            AssertDocumentNodeMatchesBaseline(generated.CodeDocument);
+            AssertCSharpDocumentMatchesBaseline(generated.CodeDocument);
+
+            var diagnostic = Assert.Single(generated.Diagnostics);
+            Assert.Same(ComponentDiagnosticFactory.GenericComponentTypeInferenceUnderspecified.Id, diagnostic.Id);
+        }
+
```

Comment:
Will it make sense to have a test that covers multiple nested levels of type inference?

Something like
```razor
<Parent>
   <InferredChild>
     <SomeOtherInferenceRoot>
        <AnotherInferredChild></AnotherInferredChild>
     </SomeOtherInferenceRoot>
   </InferredChild>
</Parent>
```

Just to exercise a bit more the code generation logic

Reply:
Also, along the lines of this. We probably also want a test for this https://github.com/dotnet/aspnetcore/pull/29767/files#r569414863

```
<InferenceRoot>
   <Child></Child>
</InferenceRoot>
<InferenceRoot>
   <Child></Child>
</InferenceRoot>
```

Reply:
These are now covered by the newly-added `ChildComponent_Generic_TypeInference_MultiLayerOverrideCascade` (and partially by `ChildComponent_Generic_TypeInference_MultiLayerCascaded` which was there before).